metaflow/plugins/kubernetes/kubernetes_decorator.py

```diff
@@ -93,9 +99,27 @@ def __init__(self, attributes=None, statically_defined=False):
             self.attributes["service_account"] = KUBERNETES_SERVICE_ACCOUNT
         if not self.attributes["gpu_vendor"]:
             self.attributes["gpu_vendor"] = KUBERNETES_GPU_VENDOR
+        if not self.attributes["node_selector"] and KUBERNETES_NODE_SELECTOR:
+            self.attributes["node_selector"] = KUBERNETES_NODE_SELECTOR
+        if not self.attributes["tolerations"] and KUBERNETES_TOLERATIONS:
+            self.attributes["tolerations"] = json.loads(KUBERNETES_TOLERATIONS)
+
+        if isinstance(self.attributes["node_selector"], str):
+            self.attributes["node_selector"] = self.parse_node_selector(
+                self.attributes["node_selector"].split(",")
+            )
 
-        # TODO: Handle node_selector in a better manner. Currently it is special
-        #       cased in kubernetes_client.py
+        if self.attributes["tolerations"]:
+            try:
+                from kubernetes.client import V1Toleration
```

Comment:
I understand that the rationale for including this check in `_init_` is to ensure that this check is invoked for `argo-workflows` too. However, this check will fail if the user hasn't installed the python package `kubernetes` yet - which is checked in `package_init` - that check should technically happen before the check for tolerations.

Reply:
It is checked in `kubernetes_cli`.
The idea is that the check is invoked only if required, which means the python package `kubernetes` must be installed. I think the order of the execution doesn't matter.
If `kubernetes` is not available, `self.attributes["tolerations"]` is not being used, then the check is not required.
Does it make sense to you?

Reply:
When the user pip installs metaflow, we don't install Kubernetes python package. It's only when the user starts executing a flow that involves `@kubernetes` or argo - we throw a nice warning asking them to install the python package. Now, if that first flow has `tolerations` defined, then the user will instead get an error saying `no module named Kubernetes`

Reply:
That import is inside a `try` block with
```
except (NameError, ImportError):
  pass
```
It should not raise any errors related to the missing module, is it correct?

Reply:
Yes - but it's the round about way this check is implemented which is my concern. We can ship this and come back to clean it up.